spring-cloud-task-core/src/main/java/org/springframework/cloud/task/repository/support/TaskExecutionDaoFactoryBean.java

```diff
@@ -0,0 +1,141 @@
+
+	private ConfigurableApplicationContext context;
+
+	private TaskExecutionDao dao = null;
+
+	private String dataSourceName;
+
+	private String tablePrefix = DEFAULT_TABLE_PREFIX;
+
+	/**
+	 * Default constructor will result in a Map based TaskExecutionDao.  <b>This is only
+	 * intended for testing purposes.</b>
+	 */
+	public TaskExecutionDaoFactoryBean() {
+	}
+
+	/**
+	 * ApplicationContext provided will be used to obtain the appropriate
+	 * {@link DataSource}.
+	 *
+	 * @param context context for this application
+	 */
+	public TaskExecutionDaoFactoryBean(ConfigurableApplicationContext context) {
+		Assert.notNull(context, "An ApplicationContext is required");
+
+		this.context = context;
+	}
+
+	@Override
+	public TaskExecutionDao getObject() throws Exception {
+		if(this.dao == null) {
+			if(this.context != null) {
+				if (StringUtils.hasText(this.dataSourceName)) {
+					if(!this.context.containsBean(this.dataSourceName)) {
+						throw new IllegalArgumentException("The configured dataSourceName is not available in the current context");
+					}
+
+					DataSource dataSource = (DataSource) this.context.getBean(this.dataSourceName);
+					buildTaskExecutionDao(dataSource);
+				}
+				else if (this.context.getBeanNamesForType(DataSource.class).length == 1) {
+					DataSource dataSource = this.context.getBean(DataSource.class);
+					buildTaskExecutionDao(dataSource);
+
+				}
+				else {
+					this.dao = new MapTaskExecutionDao();
+				}
+			}
+			else {
+				this.dao = new MapTaskExecutionDao();
+			}
+		}
+
+		return this.dao;
+	}
+
+	@Override
+	public Class<?> getObjectType() {
+		return TaskExecutionDao.class;
+	}
+
+	@Override
+	public boolean isSingleton() {
+		return true;
+	}
+
+	/**
+	 * Identifies the {@link DataSource} to be used if one is to be used.  By default, the
+	 * name is not specified and it is assumed that only one DataSource exists within the
+	 * context.
+	 *
+	 * @param dataSourceName bean id for the DataSource to be used.
+	 */
+	public void setDataSourceName(String dataSourceName) {
+		this.dataSourceName = dataSourceName;
+	}
+
+	/**
+	 * Indicates a prefix for all of the task repository's tables if the jdbc option is
+	 * used.
+	 *
+	 * @param tablePrefix the string prefix for the task table names
+	 */
+	public void setTablePrefix(String tablePrefix) {
+		this.tablePrefix = tablePrefix;
+	}
+
+	private void buildTaskExecutionDao(DataSource dataSource) {
+		DataFieldMaxValueIncrementerFactory incrementerFactory = new DefaultDataFieldMaxValueIncrementerFactory(dataSource);
+		this.dao = new JdbcTaskExecutionDao(dataSource);
+		String databaseType;
+		try {
+			databaseType = org.springframework.batch.support.DatabaseType.fromMetaData(dataSource).name();
+		}
+		catch (MetaDataAccessException e) {
+			throw new IllegalStateException(e);
+		}
+		((JdbcTaskExecutionDao) this.dao).setTaskIncrementer(incrementerFactory.getIncrementer(databaseType, this.tablePrefix + "SEQ"));
```

Comment:
Is there a benefit to that (I usually try to limit what's in there to only what can cause the exception)?